x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java

```diff
@@ -204,6 +215,92 @@ public static String exceptionToString(Exception ex) {
         }, ToXContent.EMPTY_PARAMS);
     }
 
+    private static long numPolicySnapshotsInProgress(String policyId, SnapshotsInProgress snapshots) {
+        long numInProgress = 0;
+        for (final List<SnapshotsInProgress.Entry> entriesForRepo : snapshots.entriesByRepo()) {
+            numInProgress += entriesForRepo.stream().map(SnapshotsInProgress.Entry::userMetadata)
+                .filter(Objects::nonNull)
+                .map(meta -> meta.get(SnapshotsService.POLICY_ID_METADATA_FIELD))
+                .filter(policyId::equals)
+                .count();
+        }
+        return numInProgress;
+    }
+
+    private static class PreRegisterSLMRun extends ClusterStateUpdateTask {
+        private final String policyName;
+        private final String snapshotName;
+
+        PreRegisterSLMRun(String policyName, String snapshotName) {
+            this.policyName = policyName;
+            this.snapshotName = snapshotName;
+        }
+
+        @Override
+        public ClusterState execute(ClusterState currentState) throws Exception {
+            SnapshotLifecycleMetadata snapMeta = currentState.metadata().custom(SnapshotLifecycleMetadata.TYPE);
+
+            assert snapMeta != null : "this should never be called while the snapshot lifecycle cluster metadata is null";
+            if (snapMeta == null) {
+                logger.error(
+                    "failed to pre-register snapshot [{}] in policy [{}]: snapshot lifecycle metadata is null",
+                    snapshotName,
+                    policyName
+                );
+                return currentState;
+            }
+
+            Map<String, SnapshotLifecyclePolicyMetadata> snapLifecycles = new HashMap<>(snapMeta.getSnapshotConfigurations());
+            SnapshotLifecyclePolicyMetadata policyMetadata = snapLifecycles.get(policyName);
+            if (policyMetadata == null) {
+                logger.warn(
+                    "failed to pre-register snapshot [{}] in policy [{}]: policy not found",
+                    snapshotName,
+                    policyName
+                );
+                return currentState;
+            }
+
+            long preRegisteredRuns = policyMetadata.getPreRegisteredRuns();
+            SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
+            final long numSnapshotsRunning = snapshots == null ? 0 : numPolicySnapshotsInProgress(policyName, snapshots);
+
+            SnapshotLifecyclePolicyMetadata.Builder newPolicyMetadata = SnapshotLifecyclePolicyMetadata.builder(policyMetadata);
+            if (preRegisteredRuns > numSnapshotsRunning) {
+                final long unrecordedFailures = preRegisteredRuns - numSnapshotsRunning;
```

Comment:
Counting the number of currently running snapshots and using this to update preRegisteredRuns worries me a bit. A simpler alternative would be the following:  If there are concurrently running snapshots for the same policy, just don't update `invocationsSinceLastSuccess`. I _think_ that snapshots can only run concurrently due to a manual rather than scheduled run. It does not seem terrible to just bail out in this case and not get the benefits of this PR.